src/components/users/controllers.tsx

```diff
@@ -95,3 +106,208 @@ export async function getUser(
     ),
   };
 }
+
+export async function resetPasswordRequestToken(ctx: IContext, _params: IParameters): Promise<IResponse> {
+  const template = new Template(ctx.viewContext, 'Request password reset');
+
+  return await Promise.resolve({
+    body: template.render(<PasswordResetRequest csrf={ctx.viewContext.csrf} />),
+  });
+}
+
+export async function resetPasswordObtainToken(
+  ctx: IContext, params: IParameters, body: IPasswordResetBody,
+): Promise<IResponse> {
+  const VALID_EMAIL = /^[^@]*@[^@]*$/;
+  const email = (/* istanbul ignore next */ body.email || '').toLowerCase();
+  const template = new Template(ctx.viewContext, 'Request password reset');
+
+  if (!email || !VALID_EMAIL.test(email)) {
+    template.title = 'Error: Request password reset';
+
+    return {
+      body: template.render(<PasswordResetRequest csrf={ctx.viewContext.csrf} invalidEmail={true} />),
+      status: 400,
+    };
+  }
+
+  const uaa = new UAAClient({
+    apiEndpoint: ctx.app.uaaAPI,
+    clientCredentials: {
+      clientID: ctx.app.oauthClientID,
+      clientSecret: ctx.app.oauthClientSecret,
+    },
+  });
+
+  const accounts = new AccountsClient({
+    apiEndpoint: ctx.app.accountsAPI,
+    logger: ctx.app.logger,
+    secret: ctx.app.accountsSecret,
+  });
+
+  const accountsUser = await accounts.getUserByEmail(email);
+  if (!accountsUser) {
+    // 100ms delay to stop people iterating through users
+    await new Promise(finish => setTimeout(finish, 100));
+
+    template.title = 'Error: User not found';
```

Comment:
I'm not a fan of this error being presented. This has a potential for an attacker to recognise and perform targeted attacks. Delay is good for mass snooping through users, but not targeted.

Just saying, no need to act :)

Reply:
A compromise here would be

If the user is not found, sleep a long time (eg 500ms)

If the user is found, but using SSO, send them an email saying they are using SSO?

Reply:
Yeah, previously there was a comment here with a potential to send out an educational email. It wasn't implemented though, as I was hoping the next screen would be informative enough for the users to "remember" they're using SSO.

Reply:
I am going to make the delay more aggressive - 400ms - in lieu of sending an email